tombstone_gc.hh

```diff
@@ -61,56 +61,92 @@ struct range_repair_time {
 };
 
 class tombstone_gc_state {
+public:
+    friend class tombstone_gc_before_getter;
+
+private:
     gc_time_min_source _gc_min_source;
     per_table_history_maps* _reconcile_history_maps;
-    [[nodiscard]] gc_clock::time_point check_min(schema_ptr, gc_clock::time_point) const;
 
-    [[nodiscard]] repair_history_map_ptr get_repair_history_for_table(const table_id& id) const;
     [[nodiscard]] repair_history_map_ptr get_or_create_repair_history_for_table(const table_id& id);
 
-    [[nodiscard]] seastar::lw_shared_ptr<gc_clock::time_point> get_group0_gc_time() const;
     [[nodiscard]] seastar::lw_shared_ptr<gc_clock::time_point> get_or_create_group0_gc_time();
 
-    [[nodiscard]] gc_clock::time_point get_gc_before_for_group0(schema_ptr s) const;
-
 private:
     std::unordered_map<table_id, utils::chunked_vector<range_repair_time>> _pending_updates;
 
 public:
     tombstone_gc_state() = delete;
-    explicit tombstone_gc_state(per_table_history_maps* maps) noexcept : _reconcile_history_maps(maps) {}
 
-    explicit operator bool() const noexcept {
-        return _reconcile_history_maps != nullptr;
-    }
+    // No move or copy, internals are referenced by tombstone_gc_before_getter
+    tombstone_gc_state(const tombstone_gc_state&) = delete;
+    tombstone_gc_state(tombstone_gc_state&&) = delete;
+
+    explicit tombstone_gc_state(per_table_history_maps* maps) noexcept : _reconcile_history_maps(maps) {}
 
     void set_gc_time_min_source(gc_time_min_source src) {
         _gc_min_source = std::move(src);
     }
 
-    // Returns true if it's cheap to retrieve gc_before, e.g. the mode will not require accessing a system table.
-    [[nodiscard]] bool cheap_to_get_gc_before(const schema& s) const noexcept;
-
     void drop_repair_history_for_table(const table_id& id);
 
+    void update_repair_time(table_id id, const dht::token_range& range, gc_clock::time_point repair_time);
+    void update_group0_refresh_time(gc_clock::time_point refresh_time);
+
+    // returns a tombstone_gc_state copy with the commitlog check disabled (i.e.) without _gc_min_source.
+    [[nodiscard]] tombstone_gc_state with_commitlog_check_disabled() const { return tombstone_gc_state(_reconcile_history_maps); }
+
+    void insert_pending_repair_time_update(table_id id, const dht::token_range& range, gc_clock::time_point repair_time, shard_id shard);
+    future<> flush_pending_repair_time_update(replica::database& db);
+};
+
+class tombstone_gc_before_getter {
```

Comment:
I know why you want to introduce this class. It is basically tombstone_gc_state + rf.  I am thinking if  renaming tombstone_gc_state to tombstone_gc_state_impl and renaming tombstone_gc_before_getter to tombstone_gc_before would reduce the massive changes.  We can make tombstone_gc_state cheap to copy too. The only thing that is not cheap to copy is _pending_updates.

Reply:
Renaming is needed either way. I favor the current approach because I already did all the refactoring work and I think the end result is quite nice.